runtime/src/mm/cpp/ObjectOps.cpp

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+#include "ObjectOps.hpp"
+
+#include "Common.h"
+#include "ThreadData.hpp"
+
+using namespace kotlin;
+
+// TODO: Memory barriers.
+
+ALWAYS_INLINE void mm::SetStackRef(ObjHeader** location, ObjHeader* value) noexcept {
+    *location = value;
+}
+
+ALWAYS_INLINE void mm::SetHeapRef(ObjHeader** location, ObjHeader* value) noexcept {
+    *location = value;
+}
+
+#pragma clang diagnostic push
+// On 32-bit android arm clang warns of significant performance penalty because of large
+// atomic operations. TODO: Consider using alternative ways of ordering memory operations if they
+// turn out to be more efficient on these platforms.
+#pragma clang diagnostic ignored "-Watomic-alignment"
+
+ALWAYS_INLINE void mm::SetHeapRefAtomic(ObjHeader** location, ObjHeader* value) noexcept {
+    __atomic_store_n(location, value, __ATOMIC_RELEASE);
+}
+
+ALWAYS_INLINE OBJ_GETTER(mm::ReadHeapRefAtomic, ObjHeader** location) noexcept {
+    __atomic_load(location, OBJ_RESULT, __ATOMIC_ACQUIRE);
```

Comment:
I'm not sure this `__atomic_load` is actually _atomic_.
I would expect it to atomically load from `location` and _then_ store to `OBJ_RESULT`, which doesn't exclude the possibility to observe the state between these two actions.

Reply:
If this is true, then it makes some sense to keep the implementation simple (the previous version), with a comment like one in `AllocateObject`.

Reply:
> I would expect it to atomically load from location and then store to OBJ_RESULT

Yep, you're right: https://github.com/llvm/llvm-project/blob/00530dee5d1295dc20ebafdd9a8a79662f41513e/compiler-rt/lib/builtins/atomic.c#L174

Reply:
> it makes some sense to keep the implementation simple (the previous version)

I think it'll have to stay `OBJ_GETTER`, because the implementation supporting signals will have to have this signature.

Reply:
Yes, of course.
I meant that `__atomic_load` here is somewhat confusing (because it is not atomic 🤷‍♂️).

So, if it really makes the implementation simpler and more understandable, it is ok to keep it but add a comment regarding its atomicity. If not, then replacing with real atomics makes a lot of sense.

Similar for `__atomic_compare_exchange_n` below.

Reply:
> I meant that __atomic_load here is somewhat confusing (because it is not atomic 🤷‍♂️).
> So, if it really makes the implementation simpler and more understandable, it is ok to keep it but add a comment regarding its atomicity. If not, then replacing with real atomics makes a lot of sense.

Makes sense, will do.

> Similar for __atomic_compare_exchange_n below.

Here, however, I think the implementation is fine: we write `expected` onto a stack, so it's definitely kept in the rootset; if CAS succeeds, `actual` is still in the rootset (because it equals `expected`, and `expected` is sitting on a stack); if CAS fails, `actual` is kept at the `location` and (eventually) copied onto a stack. So, it all relies on `value` and `*location` being kept alive.

Reply:
> if CAS fails, `actual` is kept at the `location` and (eventually) copied onto a stack.

And before it is copied, another thread might change the `*location`.
So I don't see how it is different from the load: if CAS fails, it acts exactly like load, and it is not atomic exactly the same way.

Reply:
Missed that, thanks! I'm keeping `__atomic_*` for consistency, though (and because it'll allow to tweak strong/weak guarantees + memory ordering; and in the future I'm thinking of using something like https://en.cppreference.com/w/cpp/atomic/atomic_ref instead of intrinsics anyway)